### PR TITLE
Handle untyped interfaces in `_check`

### PR DESCRIPTION
**Pull request by maxmynter**
Guard match checks of types in function signature and supplied via the `params` kwarg in the `benchmark` decorators `_check` method against untyped function signatures. 

If untyped, we do not check for matching types. We also log a warning to the `stdout` of the `logger` indicating the untyped parameter and function name.

### CODE REVIEW

File: src/nnbench/runner.py [hunk 1]
```diff
@@ -23,6 +23,12 @@ def _check(params: dict[str, Any], benchmarks: list[Benchmark]) -> None:
     for bm in benchmarks:
         for name, param in inspect.signature(bm.fn).parameters.items():
             param_type = param.annotation
+
+            if param.annotation == inspect.Parameter.empty:
+                logger.warn(
+                    f"Warning: Parameter {name!r} in benchmark '{bm.fn.__name__}' is untyped."
+                )
```
Comment (nicholasjng) [inline src/nnbench/runner.py:30]: I think this should be a debug log - it's more of a convenience for the user if they want to understand why something is failing, and we will always and unconditionally advertise typing benchmarks completely.

```suggestion
            if param_type == inspect.Parameter.empty:
                logger.debug(
                    f"Found untyped parameter {name!r} in benchmark {bm.fn.__name__!r}."
                )
```

File: src/nnbench/runner.py [hunk 2]
```diff
@@ -32,7 +38,9 @@ def _check(params: dict[str, Any], benchmarks: list[Benchmark]) -> None:
     for name, param in benchmark_interface.items():
         if name not in param_types and param.default == inspect.Parameter.empty:
             raise ValueError(f"missing value for required parameter {name!r}")
-        if not issubclass(param_types[name], param.annotation):
+        if not param.annotation == inspect.Parameter.empty and not issubclass(
+            param_types[name], param.annotation
+        ):  # only check type match if type supplied
```
Comment (nicholasjng) [inline src/nnbench/runner.py:43]:
```suggestion
        # skip the check if benchmark interface is untyped
        if param.annotation == inspect.Parameter.empty:
            continue
        if not issubclass(param_types[name], param.annotation):
```

**Review by nicholasjng — COMMENTED** [review]
With this change, the following should be a failure (non-unique types):

```py
import nnbench

@nnbench.benchmark
def increment(value):
    return value + 1


@nnbench.benchmark
def double(value: int) -> int:
    return value * 2
```

Is this expected? Are we fine with this?

(We should check that the resulting type in the error message is still sensible, though. Is it `<class 'empty'>` by any chance?)

**Comment by maxmynter** [comment]
> (We should check that the resulting type in the error message is still sensible, though. Is it `<class 'empty'>` by any chance?)

It is `<class 'inspect._empty'>.

**Comment by nicholasjng** [comment]
> > (We should check that the resulting type in the error message is still sensible, though. Is it `<class 'empty'>` by any chance?)
> 
> It is `<class 'inspect._empty'>`.

It's fine for now. We can think about making it show up as `<untyped>` or something, but that requires some more effort, and would only be relevant in the case of an error, i.e. just before raising the exception.

**Comment by maxmynter** [comment]
i would argue that your example above should throw an error. Untyped and typed might very well be non-unique types. We cannot handle different types for the same variable name and if untyped cannot ensure it is the right one. So it's either on the user to keep everything untyped and ensure the right value is passed. Or the absence of the type is an oversight in which case the error is a welcome reminder.

File: src/nnbench/runner.py [hunk 3]
```diff
@@ -28,15 +28,25 @@ def _check(params: dict[str, Any], benchmarks: list[Benchmark]) -> None:
                 )
 
             param_type = param.annotation
+
+            if param.annotation == inspect.Parameter.empty:
```
Comment (nicholasjng) [inline src/nnbench/runner.py:32]:
```suggestion
            if param_type == inspect.Parameter.empty:
```

**Review by nicholasjng — APPROVED** [review]

**Pull request merged**